applications/ShapeOptimizationApplication/python_scripts/mapping/sliding_vertex_morphing_mapper.py

```diff
@@ -0,0 +1,393 @@
+
+class SlidingVertexMorphingMapper():
+    """
+    The SlidingVertexMorphingMapper extends the standard Vertex Morphing approach
+    by restricting the nodal shape update of a specified model part. The nodes of this model 
+    part have to lie on a predefined background mesh and are only allowed to float/slide on it.
+    SlidingVertexMorphingMapper uses similar techniques as directional damping and
+    InPlaneVertexMorphingMapper.
+
+    Limitations:
+    - Damping can only be used if all cartesian directions (x,y,z) are damped.
+    - The projection of gradients on the surface normals has to be deactivated.
+    """
+
+    def __init__(self, origin_model_part, destination_model_part, settings):
+        from KratosMultiphysics.ShapeOptimizationApplication import mapper_factory
+        if not SpacialMapperFactory:
+            raise Exception("SlidingVertexMorphingMapper: MappingApplication is required!")
+
+        self.settings = settings
+        self.origin_model_part = origin_model_part
+        self.destination_model_part = destination_model_part
+
+        extracted_vm_settings = settings.Clone()
+        extracted_vm_settings["sliding_morphing"].SetBool(False)
+        extracted_vm_settings.RemoveValue("sliding_morphing_settings")
+        self.vm_mapper = mapper_factory.CreateMapper(origin_model_part, destination_model_part, extracted_vm_settings)
+
+        in_plane_settings = self.settings["sliding_morphing_settings"]
+        in_plane_settings.RecursivelyValidateAndAssignDefaults(self.GetDefaultInPlaneSettings())
+
+        self._background_model = KM.Model()
+        if in_plane_settings["model_import_settings"]["input_type"].GetString() in ["mdpa", "vrml", "wrl"]:
+            background_main_mesh = self._background_model.CreateModelPart("background_mesh")
+            background_main_mesh.ProcessInfo.SetValue(KM.DOMAIN_SIZE, 3)
+            background_main_mesh.AddNodalSolutionStepVariable(KM.NORMAL)
+            background_main_mesh.AddNodalSolutionStepVariable(KSO.NORMALIZED_SURFACE_NORMAL)
+            background_main_mesh.AddNodalSolutionStepVariable(KSO.BACKGROUND_COORDINATE)
+        else:
+            raise Exception("Other model part input options are not yet implemented.")
+
+        self.background_mesh = None  # created in Initialize
+        self.spacial_mapper = None   # created in Initialize
+        self.sliding_mesh = None     # created in Initialize
+
+    @classmethod
+    def GetDefaultInPlaneSettings(cls):
+        return KM.Parameters("""{
+            "model_import_settings" : {
+                "input_type" : "mdpa",
+                "input_filename" : "UNKNOWN_NAME"
+            },
+            "sliding_sub_model_part_name" : "",
+            "background_sub_model_part_name" : "",
+            "spacial_mapper_settings" : {
+                "mapper_type": "nearest_element",
+                "echo_level" : 0
+            }
+        }""")
+
+    def GetStructuralSimilaritySetttings(self):
```

Comment:
```suggestion
    def GetStructuralSimilaritySettings(self):
```
typo, please search and replace